### DISCUSSION TITLE
WISHLIST: Server side - now open!

### DISCUSSION THREAD

**Discussion by adrianhall** [Azure/azure-mobile-apps#851] [General]
Hey folks,

I have just checked in the final bit of code for the v8 server.  This means I am opening up the v8 server for "wishlist" items.  Want "GeographyPoint" to be supported?  Maybe, you want to see a per-PR build for testing?  

Note that this is SERVER SIDE ONLY at this time.  I'm not ready for wish list items for the client side yet - there is a couple of months worth of work before I get to that.

How to file a wish list item:

1. Open an issue.
2. Title it "V8 WishList: <something>"
3. Make the description very specific.   If you want to be able to write a bit of code, then make sure you include the code.
4. Submit the issue.

I'll then tag it and work on it over the next month or so.

**Comment by amertlich** [comment]
My guess is this is outside the scope of feasible wish list items but I'll throw it out there anyway ;)

I would love if a zipped/compressed SQLite database file could be transacted both as a PUSH and as a PULL payload (Packfile).  We're considering doing this ourselves (in v6) but would be amazing if it was more officially supported. Basically it allows bulk transfer upload/download of a changeset (that may even contain changes from multiple different tables) in a single file. 

On the push side -- rather than pushing up the operations queue one at a time -- we just extract the pending operations queue and associated items into their own SQLite database file and upload that so it's processed as a job in the background and notifies on completion.

On the pull side -- we check if the pending incremental sync would be more efficiently handled as a blob/Packfile (like what Git does)

Let me know if you want me to submit an issue ;)

**Reply by adrianhall** [reply]
Sounds like a reasonable idea.  Depends if I can construct and deconstruct a packfile on the server side.

My thinking on this is that the client has to request it.  If we don't do that, then the client has to guess what it is going to get - it's no longer deterministic.  I'm thinking that if ?__pack=sqlite is specified (and $skip,$top are NOT specified) then a pack file is produced.  

I believe this is two separate issues.  I also want to think of what happens when the operations queue being submitted as a packfile or blob has conflicts in it?  Does the server apply an algorithm; stop on first conflict; etc.  Also, what's the response for each case?

It needs some thought put into the protocol.

**Reply by fileman** [reply]
Server can return to the client the sqlite file with only conflict and end user choose what to do, then return to the server, repeat this until no more conflict, finally return to the client the complete sqlite file to the end user... a little more complex

**Reply by adrianhall** [reply]
That’s not going to work.  The server side database is NOT SQLite in a generalized case.  Also, copying the SQLite database over doesn’t provide the necessary conflict capabilities we are after.  

I’m thinking that implementing a bulk processor endpoint is a good idea.  Basically, this would be an async operation (so you post and get a 202 Accepted back with a location to check for results).  The payload would be a JSON array of objects (operationid, method, endpoint, content, options).  When you GET the location you get back either a 102 Processing (perhaps with a payload to indicate what has been processed) or a 200 OK with the a JSON array of objects (operationid, status code, content).

The options would include when to stop processing (e.g. on first error, continue processing other tables, or process all operations), a client ID (that would be used for ensuring that bulk operation results can be retrieved per client when/if the client restarts) and a conflict resolution mechanism (client wins, server wins, latest wins, error, etc.)

I think this allows maximal flexibility to implement the appropriate bulk processing on the client side.   This still does not allow for transactions, and it loses any ability to track where the bulk processor is in the processing to drive a progress bar, for instance.  It also requires that you have a bit of insight into your data resiliency requirements to decide whether to use it.  Nothing comes for free and a bulk processor increases the chances of a mismatch between the client side and server side.

I’m also thinking of providing a “deletion sync” capability on a table.  Basically, you post a list of IDs to the table (with a specific header) and get back the ones that are not in the (client view of the) database.  This will allow the client to do a clean-up of the table when soft-delete is not being used.

These are just thoughts at this point - more design work is needed.

**Reply by amertlich** [reply]
I see your point to avoid the SQLite packfile format especially because both server and client side aren't guaranteed to be SQLite. Having said that, it does have a secondary benefit of being able to "clone" a dataset as the starting point for your local offline store client side without having to convert and upsert a JSON payload into the local offline store (i.e it would be much faster in the "clone" use case especially for larger datasets). I agree, since AMA isn't opinionated enough to assume SQLite as the sole local store type, your approach is probably better.

But I'd also highlight that your latest comments seem to be focusing more on the PUSH side of things (which I'd say is the more valuable feature) rather than the PULL side of things. Git uses packfiles on both ends, but you may be right that there's good reasons to treat these as separate features.

--

Concerning the bulk processor endpoint design, I'd hope to have server-side hooks for customizable conflict resolution. I like having the options to indicate the automated conflict handling modes you mention, but some of our conflict resolution is a bit more sophisticated (albeit still mostly automatic/deterministic).

Would this new bulk endpoint functionality be built into each TableController or would it be its own separate APIController? Could we handle the conflicts server side by overriding a method on the strongly typed table controller? Or would it be better handled by overriding a method on the separate bulk processor APIController directly so that the sequence of operations regardless of table could be maintained?

If the bulk endpoint aligns closely with the `TableOperationError` API, then the conflict resolution we are currently doing client side could also be re-used server side for quicker bulk handling of automatic conflict resolution.  

We have edge cases where conflicts are unresolvable through automated methods, requiring end-user confirmation. For instance, resurrecting a remotely deleted object that was locally updated more recently has implications for related data in our schema. Using 102 Processing and 200 OK makes sense, but we may also need 409 for unresolvable conflicts or 428 in specific cases. Purging/acknowledging these error states could be similar to client-side procedures, so storing errors in the server database until client acknowledgment seems necessary.

The configuration options you describe as whether or not this is a "stop on 1st conflict" vs  "continue and batch errors" probably require some discussion on the design too. But I imagine this is similar to the client side error handling that exists today where push `TableOperationError`s appear batched when you're using parallel options but are effectively fail-on-first if not.

Appreciate the thoughts gone into this! I hope we can arrive at a simple and intuitive solution despite how complex it could potentially be because I think this would solve some of the largest problems I see with v6.

**Reply by adrianhall** [reply]
I’m leaning towards a separate API controller for bulk processing, but I’m open to other thoughts.

Unless you do it first, I’ll start some (a few) WISHLIST issues so we can discuss design separately for each one.
